Split align related css property parsing out into its own file

weinig opened the pull request:
#### d489dc4fc3730b5fc9c554d02b32ef2828088e4d
<pre>
Split align related css property parsing out into its own file
<a href="https://bugs.webkit.org/show_bug.cgi?id=279299">https://bugs.webkit.org/show_bug.cgi?id=279299</a>

Reviewed by Darin Adler.

Splits property consumers for CSS Align properties into their own file and refactor
them a bit to hide the implementation details.

We now have explicit functions for each property, named appropriately to match the
property. The implementations remain the same, still calling to shared consumers
and passing predicates to refine behavior, but from the outside, things are much
more clear.

The shorthands were also refactored and now all use a shared consumeAlignShorthand
consumer function.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Align.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Align.h: Added.
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/css/process-css-properties.py:

Canonical link: <a href="https://commits.webkit.org/283317@main">https://commits.webkit.org/283317@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9d903948a36c4e5f7c44a76a62f3511a18afcb8b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65967 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45340 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18586 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69996 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16576 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68085 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53139 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16858 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52944 "Passed tests") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/11528 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69034 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41844 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57110 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33579 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38515 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14487 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15452 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60386 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14834 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71699 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9922 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14246 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60260 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9954 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57172 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/60550 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14539 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8199 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1834 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41148 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42224 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43407 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41968 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2332,29 +2333,37 @@ bool CSSPropertyParser::consumeGridShorthand(bool important)
-    if (!justifyContentValue)
-        return false;
-    if (!m_range.atEnd())
+        RefPtr justifyContentValue = alignContentValue.copyRef();
```
Why isn't this `Ref` instead of `RefPtr`? Why is explicit `copyRef` needed?

weinig replied:
I don't have a good justification for the RefPtr here other than it made it so both sets of addProperty() calls were identical. As for the copyRef(), that is just a mistake, I don't think it is needed.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2364,20 +2373,28 @@ bool CSSPropertyParser::consumePlaceContentShorthand(bool important)
-    if (!justifyItemsValue)
-        return false;
+    if (m_range.atEnd()) {
+        RefPtr justifyItemsValue = alignItemsValue.copyRef();
```
Why isn't this `Ref` instead of `RefPtr`? Why is explicit `copyRef` needed?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2387,27 +2404,65 @@ bool CSSPropertyParser::consumePlaceItemsShorthand(bool important)
-    if (!justifySelfValue)
-        return false;
+    if (m_range.atEnd()) {
+        RefPtr justifySelfValue = alignSelfValue.copyRef();
```
Why isn't this `Ref` instead of `RefPtr`? Why is explicit `copyRef` needed?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2387,27 +2404,65 @@ bool CSSPropertyParser::consumePlaceItemsShorthand(bool important)
+        return false;
+
+    if (m_range.atEnd()) {
+        RefPtr columnGap = rowGap.copyRef();
```
Why isn't this `Ref` instead of `RefPtr`? Why is explicit `copyRef` needed?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2332,29 +2333,37 @@ bool CSSPropertyParser::consumeGridShorthand(bool important)
-        return false;
-    if (!m_range.atEnd())
+        RefPtr justifyContentValue = alignContentValue.copyRef();
+        addProperty(CSSPropertyAlignContent, CSSPropertyPlaceContent, alignContentValue.releaseNonNull(), important);
```
Maybe we should add a helper that sets two properties to the same value to clean up the idiom in all of these clauses. Since there are 4 of them that seems like enough to justify it.

weinig replied:
I think that is a great idea! I have a few questions about how to get it right:

1. Is consume2ValueShorthand() already doing this? (as much time as I have spent on property parsing, I have spent little on the shorthands).

2. If we use the pattern I used exactly, we could have something like:

```c++
bool CSSPropertyParser::consumePairedShorthand(const StylePropertyShorthand& shorthand, bool important)
{
    ASSERT(shorthand.length() == 2);
    const CSSPropertyID* longhands = shorthand.properties();

    RefPtr prop1 = parseSingleValue(longhands[0], shorthand.id());
    if (!prop1)
        return false;

    if (m_range.atEnd()) {
        RefPtr prop2 = prop1;
        addProperty(longhands[0], shorthand.id(), prop1.releaseNonNull(), important);
        addProperty(longhands[1], shorthand.id(), prop2.releaseNonNull(), important);
        return true;
    }

    RefPtr prop2 = parseSingleValue(longhands[1], shorthand.id());
    if (!prop2 || !m_range.atEnd())
        return false;

    addProperty(longhands[0], shorthand.id(), prop1.releaseNonNull(), important);
    addProperty(longhands[1], shorthand.id(), prop2.releaseNonNull(), important);
    return true;
}
```

But that would leave out `place-content` since that has an extra check.

Alternatively, we could back to the way they were doing it like:

```c++
bool CSSPropertyParser::consumePairedShorthand(const StylePropertyShorthand& shorthand, bool important)
{
    ASSERT(shorthand.length() == 2);
    const CSSPropertyID* longhands = shorthand.properties();

    auto rangeCopy = m_range;

    RefPtr prop1 = parseSingleValue(longhands[0], shorthand.id());
    if (!prop1)
        return false;

    // If there are no more tokens, that prop2 should use re-use the original range. This is the equivalent of copying and validating prop1.
    if (m_range.atEnd())
        m_range = rangeCopy;

    RefPtr prop2 = parseSingleValue(longhands[1], shorthand.id());
    if (!prop2 || !m_range.atEnd())
        return false;

    addProperty(longhands[0], shorthand.id(), prop1.releaseNonNull(), important);
    addProperty(longhands[1], shorthand.id(), prop2.releaseNonNull(), important);
    return true;
}
```

3. Have a good name for this? (`consumePairedShorthand` was a placeholder).

darinadler replied:
I like what you ended up doing. Sorry I didn’t help more with the exploration of possibilities!

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Align.cpp:
```diff
@@ -0,0 +1,216 @@
+namespace WebCore {
+namespace CSSPropertyParserHelpers {
+
+using IsPositionKeyword = bool (*)(CSSValueID);
```
You moved this and did not create it. But note that this type sounds a little bit like a boolean or an enumeration we use as a named boolean and programmers could make the mistake of using it as a boolean, which would compare against null so turn into `true` and probably not even warn. So maybe tweak the name slightly?

weinig replied:
Renamed to `PositionKeywordPredicate`.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -2254,65 +2145,31 @@ RefPtr<CSSValue> consumeSingleWebkitBackgroundSize(CSSParserTokenRange& range, c
+bool isFlexBasisIdent(CSSValueID id)
 {
-    return isSelfPositionKeyword(id) || isLeftOrRightKeyword(id);
+    switch (id) {
```
How did you decide whether to use `identMatches` or not?

weinig replied:
In this case, I was merging in validWidthOrHeightKeyword(), since it was the only use, and it used a switch, so I just kept it that way. My feeling is that a switch is more natural and easier to read than identMatches<> for long lists, but don't have a strong preference.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2331,81 +2332,33 @@ bool CSSPropertyParser::consumeGridShorthand(bool important)
-    return true;
-}
+    ASSERT(shorthand.length() == 2);
+    const CSSPropertyID* longhands = shorthand.properties();
```
```suggestion
    auto* longhands = shorthand.properties();
```

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Align.h:
```diff
@@ -0,0 +1,64 @@
+class CSSParserTokenRange;
+class CSSValue;
+
+enum CSSValueID : uint16_t;
```
Just noticed this header doesn’t use this type. So we can remove this forward declaration.

darinadler replied:
Not that we should hold up the merging for this tiny tweak.

weinig committed:
Split align related css property parsing out into its own file
https://bugs.webkit.org/show_bug.cgi?id=279299

Reviewed by Darin Adler.

Splits property consumers for CSS Align properties into their own file and refactor
them a bit to hide the implementation details.

We now have explicit functions for each property, named appropriately to match the
property. The implementations remain the same, still calling to shared consumers
and passing predicates to refine behavior, but from the outside, things are much
more clear.

The shorthands were also refactored and now all use a shared consumeAlignShorthand
consumer function.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Align.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Align.h: Added.
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/css/process-css-properties.py:

Canonical link: https://commits.webkit.org/283317@main
